src/GraphQL/Internal/Validation.hs

```diff
@@ -65,72 +71,71 @@ import qualified Data.Set as Set
 import qualified GraphQL.Internal.AST as AST
 -- Directly import things from the AST that do not need validation, so that
 -- @AST.Foo@ in a type signature implies that something hasn't been validated.
-import GraphQL.Internal.AST (Name, Alias, TypeCondition)
+import GraphQL.Internal.AST (Name, Alias, TypeCondition, Variable)
 import GraphQL.Internal.Schema (HasName(..))
+import GraphQL.Value
+  ( Value
+  , Value'
+  , ConstScalar
+  , UnresolvedVariableValue
+  , astToVariableValue
+  )
 
 -- | A valid query document.
 --
 -- Construct this using 'validate' on an 'AST.QueryDocument'.
-data QueryDocument
+data QueryDocument value
   -- | The query document contains a single anonymous operation.
-  = LoneAnonymousOperation Operation Fragments
+  = LoneAnonymousOperation (Operation value)
   -- | The query document contains multiple uniquely-named operations.
-  | MultipleOperations Operations Fragments
+  | MultipleOperations (Operations value)
   deriving (Eq, Show)
 
-data Operation
-  = Query [AST.VariableDefinition] Directives SelectionSet
-  | Mutation  [AST.VariableDefinition] Directives SelectionSet
+data Operation value
+  = Query VariableDefinitions (Directives value) (SelectionSet value)
+  | Mutation VariableDefinitions (Directives value) (SelectionSet value)
   deriving (Eq, Show)
 
+instance Functor Operation where
+  fmap f (Query vars directives selectionSet) = Query vars (fmap f directives) (map (fmap f) selectionSet)
```

Comment:
You know, they might well be. I wrote these because it helped me figure out what I was doing. Because I don't understand how `DeriveTraversable` and friends work, I left them as is.

However, I should learn! I'll write some tests (maybe won't commit them) and do some reading and update.  TODO